File: ios-app/Doompedia/Data/PackInstaller.swift
import CryptoKit
import Foundation

final class PackInstaller {
    private let store: SQLiteStore
    private let decoder = JSONDecoder()

    init(store: SQLiteStore) {
        self.store = store
    }

    func install(from directory: URL, expectedPackID: String? = nil) throws -> PackManifest {
        let manifestURL = directory.appendingPathComponent("manifest.json")
        let manifestData = try Data(contentsOf: manifestURL)
        let manifest = try decoder.decode(PackManifest.self, from: manifestData)

        if let expectedPackID, manifest.packId != expectedPackID {
            throw NSError(domain: "Doompedia", code: 41, userInfo: [NSLocalizedDescriptionKey: "Unexpected pack id \(manifest.packId)"])
        }

        for shard in manifest.shards {
            let shardURL = resolveShardURL(baseDirectory: directory, shardPath: shard.url)
            try validateChecksum(url: shardURL, expected: shard.sha256)
            try applyShard(fileURL: shardURL)
        }

        return manifest
    }

    private func applyShard(fileURL: URL) throws {
        if fileURL.pathExtension == "zst" {
            throw NSError(
                domain: "Doompedia",
                code: 42,
                userInfo: [NSLocalizedDescriptionKey: "Compressed shard install requires decoder integration"]
            )
        }

        var batch: [SeedRow] = []

        let lineReader: ((URL, (String) throws -> Void) throws -> Void) = fileURL.pathExtension == "gz"
            ? GzipLineReader.forEachLine
            : LineReader.forEachLine

        try lineReader(fileURL) { line in
            let payload = try decoder.decode(ShardRowPayload.self, from: Data(line.utf8))
            let record = payload.article
            batch.append(
                SeedRow(
                    page_id: record.page_id,
                    lang: record.lang,
                    title: record.title,
                    summary: record.summary,
                    wiki_url: record.wiki_url,
                    topic_key: record.topic_key,
                    quality_score: record.quality_score,
                    is_disambiguation: record.is_disambiguation,
                    source_rev_id: record.source_rev_id,
                    updated_at: record.updated_at,
                    aliases: payload.aliases
                )
            )

            if batch.count >= 1000 {
                try store.upsertSeedRows(batch)
                batch.removeAll(keepingCapacity: true)
            }
        }

        if !batch.isEmpty {
            try store.upsertSeedRows(batch)
        }
    }

    private func validateChecksum(url: URL, expected: String) throws {
        let data = try Data(contentsOf: url)
        let digest = SHA256.hash(data: data)
        let hash = digest.map { String(format: "%02x", $0) }.joined()
        guard hash.lowercased() == expected.lowercased() else {
            throw NSError(domain: "Doompedia", code: 43, userInfo: [NSLocalizedDescriptionKey: "Checksum mismatch for \(url.lastPathComponent)"])
        }
    }

    private func resolveShardURL(baseDirectory: URL, shardPath: String) -> URL {
        let direct = baseDirectory.appendingPathComponent(shardPath)
        if FileManager.default.fileExists(atPath: direct.path) {
            return direct
        }

        return baseDirectory.appendingPathComponent(URL(fileURLWithPath: shardPath).lastPathComponent)
    }
}

File: android-app/app/src/main/java/com/feelbachelor/doompedia/data/importer/PackInstaller.kt
package com.feelbachelor.doompedia.data.importer

import androidx.room.withTransaction
import com.feelbachelor.doompedia.data.db.AliasEntity
import com.feelbachelor.doompedia.data.db.ArticleEntity
import com.feelbachelor.doompedia.data.db.WikiDatabase
import com.feelbachelor.doompedia.domain.normalizeSearch
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import kotlinx.serialization.json.Json
import java.io.File
import java.security.MessageDigest
import java.util.zip.GZIPInputStream

class PackInstaller(
    private val db: WikiDatabase,
    private val json: Json = Json { ignoreUnknownKeys = true },
) {
    suspend fun installFromDirectory(directory: File, expectedPackId: String? = null): PackManifest {
        return withContext(Dispatchers.IO) {
            val manifestFile = File(directory, "manifest.json")
            require(manifestFile.exists()) { "manifest.json not found in ${directory.path}" }
            val manifest = json.decodeFromString(PackManifest.serializer(), manifestFile.readText())
            if (expectedPackId != null) {
                require(manifest.packId == expectedPackId) {
                    "Manifest packId ${manifest.packId} does not match expected $expectedPackId"
                }
            }

            manifest.shards.forEach { shard ->
                val shardFile = resolveShardPath(directory, shard.url)
                require(shardFile.exists()) { "Missing shard file: ${shardFile.path}" }
                val digest = shardFile.sha256()
                require(digest.equals(shard.sha256, ignoreCase = true)) {
                    "Checksum mismatch for shard ${shard.id}"
                }
                applyShard(shardFile)
            }

            manifest
        }
    }

    private suspend fun applyShard(shardFile: File) {
        val inputStream = when {
            shardFile.name.endsWith(".gz") -> GZIPInputStream(shardFile.inputStream())
            shardFile.name.endsWith(".zst") -> error("zstd shards require runtime decoder integration")
            else -> shardFile.inputStream()
        }

        val articleBuffer = mutableListOf<ArticleEntity>()
        val aliasBuffer = mutableListOf<AliasEntity>()
        val pageIds = mutableListOf<Long>()

        inputStream.bufferedReader().useLines { lines ->
            lines.forEach { line ->
                if (line.isBlank()) return@forEach
                val row = json.decodeFromString(ShardRow.serializer(), line)
                val article = row.article
                articleBuffer += ArticleEntity(
                    pageId = article.page_id,
                    lang = article.lang,
                    title = article.title,
                    normalizedTitle = article.normalized_title.ifBlank { normalizeSearch(article.title) },
                    summary = article.summary,
                    wikiUrl = article.wiki_url,
                    topicKey = TopicClassifier.normalizeTopic(
                        rawTopic = article.topic_key,
                        title = article.title,
                        summary = article.summary,
                    ),
                    qualityScore = article.quality_score,
                    isDisambiguation = article.is_disambiguation,
                    sourceRevId = article.source_rev_id,
                    updatedAt = article.updated_at,
                )
                pageIds += article.page_id
                row.aliases.forEach { alias ->
                    aliasBuffer += AliasEntity(
                        pageId = article.page_id,
                        lang = article.lang,
                        alias = alias,
                        normalizedAlias = normalizeSearch(alias),
                    )
                }

                if (articleBuffer.size >= 1_000) {
                    flushBatch(articleBuffer, aliasBuffer, pageIds)
                }
            }
        }

        flushBatch(articleBuffer, aliasBuffer, pageIds)
    }

    private suspend fun flushBatch(
        articleBuffer: MutableList<ArticleEntity>,
        aliasBuffer: MutableList<AliasEntity>,
        pageIds: MutableList<Long>,
    ) {
        if (articleBuffer.isEmpty()) return
        db.withTransaction {
            val dao = db.wikiDao()
            dao.upsertArticles(articleBuffer)
            dao.deleteAliasesForPages(pageIds)
            if (aliasBuffer.isNotEmpty()) {
                dao.insertAliases(aliasBuffer)
            }
        }
        articleBuffer.clear()
        aliasBuffer.clear()
        pageIds.clear()
    }

    private fun resolveShardPath(baseDirectory: File, manifestPath: String): File {
        val candidate = File(baseDirectory, manifestPath)
        if (candidate.exists()) return candidate

        val trimmed = manifestPath.substringAfterLast('/')
        val fallback = File(baseDirectory, trimmed)
        if (fallback.exists()) return fallback

        return candidate
    }
}

private fun File.sha256(): String {
    val digest = MessageDigest.getInstance("SHA-256")
    inputStream().use { stream ->
        val buffer = ByteArray(DEFAULT_BUFFER_SIZE)
        while (true) {
            val read = stream.read(buffer)
            if (read <= 0) break
            digest.update(buffer, 0, read)
        }
    }
    return digest.digest().joinToString(separator = "") { "%02x".format(it) }
}
